standard/types.md

```diff
@@ -705,3 +714,182 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+When the nullable context is ***warnings***:
+
+The compiler uses static flow analysis to determine the null state of any reference variable. When nullable warnings are enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null*, *maybe null*, or *maybe default* and
+
+- A warning shall be generated if a reference variable that has been determined to be *maybe null*, is dereferenced.
+- The state of a reference expression is *maybe null* unless the compiler can determine one of two conditions:
+  1. The variable has been definitely assigned a non-`null` value.
+  1. The variable has been checked against `null` before dereferencing it.
+
+Some warnings are generated when the nullable context is set to *warnings*. (Examples include throwing a value that may be `null`, and dereferencing a possibly null reference.) Other warnings are generated only if the nullable context is enabled (§Nullable-Warning-Context). (Examples include converting a null literal or possible null value to a non-nullable type, possible null reference assignment, and possible null reference return.)
+
+> *Example*:
+>
+> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext3", ignoredWarnings:["CS0219"], expectedWarnings:["CS8632", "CS8602"], expectedException:"NullReferenceException"} -->
+> ```csharp
+> #nullable disable annotations
+> #nullable enable warnings
+> string? s1 = null;    // OK; ? makes s2 nullable
+> string s2 = null;     // OK; null-state of s2 is "maybe null"
+> s2 = null;            // OK; null-state of s2 is "maybe null"
+> char c1 = s2[1];      // Warning; dereference of a possible null; throws NullReferenceException
+> c1 = s2![1];          // The warning is suppressed
+> ```
+>
+> *end example*
+
+#### §Nullable-enable-Context Nullable enable
+
+When the nullable context is ***enabled***:
+
+- For any reference type `T`, the annotation `?` in `T?` makes `T?` a nullable type, whereas the unannotated `T` is non-nullable.
+- The compiler uses static flow analysis to determine the null state of any reference variable. When nullable warnings are enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null*, *maybe null*, or *maybe default* and
+  - A warning shall be generated when a possible null value is assigned to a variable of a non-nullable type.
+  - A warning shall be generated if a reference expression that has been determined to be *maybe null*, is dereferenced.
+  - The state of a non-nullable reference variable is *not null* unless the compiler can determine that the variable was assigned to an expression that is *maybe null*.
+  - The state of a nullable reference expression is *maybe null* unless the compiler can determine one of two conditions:
+    1. The variable has been definitely assigned a non-`null` value.
+    1. The variable has been checked against `null` before dereferencing it.
+
```

Comment:
Extra note from Mads: checking against null affects things even outside the branch:

```#nullable enable
string x = "";
// With this line, there's a warning on x.Length.
// Without this line, there's no warning
if (x is null) { }
Console.WriteLine(x.Length);
```

Reply:
This *could* be something that we don't make normative - but might be a helpful example of "additional implementation-defined null state tracking".

Reply:
I think it *is* something we don’t make normative.

As this collection of PRs develops we need to consider how many requirements/constraints are being place on a null analysis that itself is not being specified. My current feeling is we're doing erring on the side of specifying too much, but things may change…